### PR TITLE
Map vector types to arrays and pass them by value

### PR DESCRIPTION
**Pull request by gnzlbg**
This PR maps vector types to arrays and passes them by value (that is, they are passed to C as `[T; N]`.  This already allows defining them as a blacklisted opaque type, such that the user can provide its own definition from e.g. `std::arch`. 

A subsequent PR should map vector types to unit structs with a private member:

```rust
#[repr(align(16))] pub struct __m128([f32; 4]); 
```

to make their alignment at least be correct. This should allow transmuting `std::arch` types to these types properly. 

Once that is done, we probably want to detect the canonical vector types (e.g. `__m128`) and pull in the type from `std`/`core`::arch instead.

### CODE REVIEW

**Comment by gnzlbg** [comment]
The builds failures appear spurious. 

I'm using this branch to finally build `sleef-sys` successfully (https://github.com/gnzlbg/sleef-sys), so I would prefer to focus into getting some sort of this merged in an usable way that breaks nobody first, and ironing out other issues later.

File: src/ir/ty.rs [hunk 1]
```diff
@@ -627,6 +630,9 @@ pub enum TypeKind {
     /// template parameters.
     TemplateAlias(TypeId, Vec<TypeId>),
 
+    /// A packed vector type: element type, number of elements, vector alignment.
+    Vector(TypeId, usize, usize),
```
Comment (emilio) [inline src/ir/ty.rs]: Can we name the fields if we need to keep the alignment?

File: src/ir/ty.rs [hunk 2]
```diff
@@ -1148,12 +1154,15 @@ impl Type {
...
-                // That being said, that should be fixed eventually.
-                CXType_Vector |
+                CXType_Vector => {
+                    let inner = Item::from_ty(
+                        ty.elem_type().as_ref().unwrap(),
+                        location,
+                        None,
+                        ctx,
+                    ).expect("Not able to resolve vector element?");
+                    TypeKind::Vector(inner, ty.num_elements().unwrap(), ty.align())
```
Comment (emilio) [inline src/ir/ty.rs]: Is `ty.align()` the same as `layout().align`? If so, there's no need to store it in the vector.

File: src/ir/analysis/derive_partialeq_or_partialord.rs [hunk 3]
```diff
@@ -199,6 +199,14 @@ impl<'ctx> CannotDerivePartialEqOrPartialOrd<'ctx> {
                     return CanDerive::ArrayTooLarge;
                 }
             }
+            TypeKind::Vector(..) => {
+                // FIXME: vectors always can derive PartialEq, but they can
+                // never derive PartialOrd. 
+                trace!(
```
Comment (emilio) [inline src/ir/analysis/derive_partialeq_or_partialord.rs]: nit: This fits in one line.

File: src/ir/analysis/derive_hash.rs [hunk 4]
```diff
@@ -202,6 +202,18 @@ impl<'ctx> MonotoneFramework for CannotDeriveHash<'ctx> {
...
             }
+            TypeKind::Vector(t, len, _) => {
+                if self.cannot_derive_hash.contains(&t.into()) {
+                    trace!(
+                        "    vectors of T for which we cannot derive Hash \
+                         also cannot derive Hash"
+                    );
+                    return self.insert(id);
+                }
+                assert!(len != 0);
```
Comment (emilio) [inline src/ir/analysis/derive_hash.rs]: nit: `assert_ne!` gives a nicer error message.

File: src/codegen/mod.rs [hunk 5]
```diff
@@ -3055,6 +3056,12 @@ impl TryToRustTy for Type {
                     [ #ty ; #len ]
                 })
             }
+            TypeKind::Vector(item, len, _) => {
```
Comment (emilio) [inline src/codegen/mod.rs]: We can merge this path with the one above for now if you want.

File: src/ir/analysis/derive_partialeq_or_partialord.rs [hunk 6]
```diff
@@ -199,6 +199,14 @@ impl<'ctx> CannotDerivePartialEqOrPartialOrd<'ctx> {
                     return CanDerive::ArrayTooLarge;
                 }
             }
+            TypeKind::Vector(..) => {
+                // FIXME: vectors always can derive PartialEq, but they can
+                // never derive PartialOrd.
```
Comment (emilio) [inline src/ir/analysis/derive_partialeq_or_partialord.rs]: Maybe point to the link you sent over IRC for reference about why?

**Comment by gnzlbg** [comment]
@emilio I think I have addressed all the comments. I've removed the alignment for now (it is unused as you said), and fixed the other changes suggested.

**Comment by emilio** [comment]
@bors-servo r+

Sure, will publish as soon as it merges :)

**Comment by emilio** [comment]
0.38.0 should be up now, sorry for the lag!

**Comment by gnzlbg** [comment]
Thanks!

**Pull request merged**